test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs

```diff
@@ -571,6 +575,93 @@ public void AnomalyDetectionWithSrCnn(bool loadDataFromFile)
+                dataView = ml.Data.LoadFromTextFile(dataPath, new[] {
+                    new TextLoader.Column("Value", DataKind.Double, 0),
+                }, hasHeader: true);
+            }
+            else
+            {
+                // Generate sample series data with an anomaly
+                var data = new List<TimeSeriesDataDouble>();
+                for (int index = 0; index < 20; index++)
+                {
+                    data.Add(new TimeSeriesDataDouble { Value = 5 } );
+                }
+                data.Add(new TimeSeriesDataDouble { Value = 10 });
+                for (int index = 0; index < 5; index++)
+                {
+                    data.Add(new TimeSeriesDataDouble { Value = 5 });
+                }
+
+                // Convert data to IDataView.
+                dataView = ml.Data.LoadFromEnumerable(data);
+            }
+
+            // Setup the detection arguments
+            string outputColumnName = nameof(SrCnnAnomalyDetection.Prediction);
+            string inputColumnName = nameof(TimeSeriesDataDouble.Value);
+
+            // Do batch anomaly detection
+            var outputDataView = ml.Data.BatchDetectAnomalyBySrCnn(dataView, outputColumnName, inputColumnName,
+                threshold: 0.35, batchSize: -1, sensitivity: 90.0, mode);
+
+            // Getting the data of the newly created column as an IEnumerable of
+            // SrCnnAnomalyDetection.
+            var predictionColumn = ml.Data.CreateEnumerable<SrCnnAnomalyDetection>(
+                outputDataView, reuseRowObject: false);
+
+            int k = 0;
+            foreach (var prediction in predictionColumn)
+            {
+                switch (mode)
+                {
+                    case SrCnnDetectMode.AnomalyOnly:
+                        Assert.Equal(3, prediction.Prediction.Length);
+                        if (k == 20)
+                            Assert.Equal(1, prediction.Prediction[0]);
+                        else
+                            Assert.Equal(0, prediction.Prediction[0]);
+                        break;
+                    case SrCnnDetectMode.AnomalyAndExpectedValue:
+                        Assert.Equal(4, prediction.Prediction.Length);
+                        if (k == 20)
+                        {
+                            Assert.Equal(1, prediction.Prediction[0]);
+                            Assert.Equal("5.00", prediction.Prediction[3].ToString("0.00"));
+                        }
+                        else
+                            Assert.Equal(0, prediction.Prediction[0]);
+                        break;
+                    case SrCnnDetectMode.AnomalyAndMargin:
+                        Assert.Equal(7, prediction.Prediction.Length);
+                        if (k == 20)
+                        {
+                            Assert.Equal(1, prediction.Prediction[0]);
+                            Assert.Equal("5.00", prediction.Prediction[3].ToString("0.00"));
```

Comment:
>Equal [](start = 35, length = 5)

`Assert.Equal` should have an overload that tests equality with a `precision` argument, it may be better to use that than converting to string.

Reply:
done.

---
In reply to: [426445484](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426445484) [](ancestors = 426445484)